### PR TITLE
Deep Clustering Updates

### PR DESCRIPTION
**Pull request by collinleiber**
- Add the possibility to load autoencoders from file
- Add possibility to load custom_dataloaders from file
- Add tdqm (#81 ) so the current status of deep clustering training is better visible
- Add augmented_loss to the neural network architectures, which can be called by various deep clustering approaches
- Add function to create default dataloaders with augmentation (was already included, but only for test purposes)
- Change namings to be more general: autoencoder -> neural networks; reconstruction loss -> ssl_loss
- Renamed neural network parameter 'reusable' to 'work_on_copy' to be more precise

### CODE REVIEW

File: clustpy/deep/_data_utils.py [hunk 1]
```diff
@@ -224,6 +224,125 @@ def get_dataloader(X: np.ndarray | torch.Tensor, batch_size: int, shuffle: bool
...
+
+    Parameters
+    ----------
+    X : np.ndarray | torch.Tensor
+        the actual data set (can be np.ndarray or torch.Tensor)
+    batch_size : int
+        the batch size (default: 256)
+    conv_used : bool
+        defines whether a convolutional network will be used afterward.
+        In this case, grayscale images will be transformed to receive three color channels (default: False)
```
Comment (Lumik7) [inline clustpy/deep/_data_utils.py:298]: maybe change 
>  In this case, grayscale images will be transformed to receive three color channels (default: False)
to
 In this case, grayscale images will be transformed to receive three color channels by copying the grayscale channel three times (default: False)

File: clustpy/deep/_train_utils.py [hunk 2]
```diff
@@ -57,45 +110,49 @@ def get_trained_network(trainloader: torch.utils.data.DataLoader = None, data: n
...
-                "WARNING: embedding_size is larger than the dimensionality of the input dataset. embedding_size: {0} / input dimensionality: {1}".format(
-                    embedding_size, input_dim))
-        # Init neural network parameters
-        layers = _get_default_layers(input_dim, embedding_size)
-        neural_network = neural_network_class(layers=layers)
-    assert hasattr(neural_network,
-                   "fitted"), "Neural network has no attribute 'fitted' and is therefore not compatible. Check documentation of fitted, e.g., at clustpy.deep.autoencoders._abstract_autoencoder._AbstractAutoencoder"
-    # Save neural network to device
+    # Get neural network object
+    input_dim = torch.numel(next(iter(trainloader))[1][0])
```
Comment (Lumik7) [inline clustpy/deep/_train_utils.py:137]: could be put into a function to make this expression easier to understand

File: clustpy/deep/neural_networks/_abstract_autoencoder.py [hunk 3]
```diff
@@ -181,10 +182,42 @@ def loss(self, batch: list, loss_fn: torch.nn.modules.loss._Loss, device: torch.
...
+        -------
+        loss : (torch.Tensor, torch.Tensor, torch.Tensor)
+            the combined network loss of the sample and the augmented sample,
+            the embedded input sample,
+            the reconstructed input sample,
+            the embedded augmented sample,
+            the reconstructed augmented sample
+        """
+        assert type(batch) is list, "batch must come from a dataloader and therefore be of type list"
+        batches_orig = [batch[i] for i in range(2, len(batch), 2)]
```
Comment (Lumik7) [inline clustpy/deep/neural_networks/_abstract_autoencoder.py:213]: why do you need the list comprehension here? couldn't you just use:

`batches_orig = batch[2]`

maybe add a comment to clarify this further

File: clustpy/deep/neural_networks/_abstract_autoencoder.py [hunk 4]
```diff
@@ -181,10 +182,42 @@ def loss(self, batch: list, loss_fn: torch.nn.modules.loss._Loss, device: torch.
...
+        batches_aug = [batch[i] for i in range(1, len(batch), 2)]
+        loss_orig, embedded, reconstructed = self.loss([batch[0]] + batches_orig, ssl_loss_fn, device)
```
Comment (Lumik7) [inline clustpy/deep/neural_networks/_abstract_autoencoder.py:215]: add here a comment that batch[0] are the indices

File: clustpy/deep/enrc.py [hunk 5]
```diff
@@ -800,15 +815,16 @@ def enrc_encode_decode_batchwise_with_loss(V: torch.Tensor, centers: list, model
             embeddings.append(z_rot.detach().cpu())
             z_rot_back = _rotate_back(z_rot=z_rot, V=V)
             reconstruction = model.decode(z_rot_back)
```
Comment (Lumik7) [inline clustpy/deep/enrc.py:817]: this would not work for every model. only for autoencoders, which have the `decode` function. maybe we can leave it for now that ENRC can only be used with the abstract autoencoder class? Otherwise the code would need some restructuring. I can do that, but not sure when I will have the time.

File: clustpy/deep/enrc.py [hunk 6]
```diff
@@ -2211,11 +2237,10 @@ def reconstruct_subspace_centroids(self, subspace_index: int = 0) -> np.ndarray:
         cluster_space_centers = self.cluster_centers_[subspace_index]
         # rotate back as centers are in the V-rotated space
         centers_rot_back = np.matmul(cluster_space_centers, self.V.transpose())
-        centers_rec = self.autoencoder.decode(torch.from_numpy(centers_rot_back).float().to(self.device))
+        centers_rec = self.neural_network.decode(torch.from_numpy(centers_rot_back).float().to(self.device))
```
Comment (Lumik7) [inline clustpy/deep/enrc.py:2240]: this only works with an autoencoder as well.

**Review by Lumik7 — APPROVED** [review]
Nice work. I have mostly minor comments that should be adressed, otherwise it can be merged from my side

File: clustpy/deep/neural_networks/_abstract_autoencoder.py (hunk 3 above)
Comment (collinleiber) [reply clustpy/deep/neural_networks/_abstract_autoencoder.py:213]: added a comment

Comment (Lumik7) [reply clustpy/deep/neural_networks/_abstract_autoencoder.py:216]: I see, makes sense :)
